Fix job `error_message`

delapuente opened the pull request:
Fix #10 

When a job fails, requesting the result to the API returns the error details.

diego-plan9 commented on qiskit/providers/ibmq/ibmqjob.py:
```diff
@@ -271,8 +285,7 @@ def status(self):
             self._status = JobStatus.ERROR
-            # TODO: This seems to be an inconsistency in the API package.
-            self._api_error_msg = api_job.get('error') or api_job.get('Error')
+            self._api_error_msg = self._get_error_details()
```
This would end up implying a call to `api.get_job()`, and fetching the results, right? One of the constraints introduced previously was that that `job.status()` should explicitly _not_ consume results (ie. only call `api.get_status_job()`, as in some premium configurations, the results can only be read once, and are deleted after the first read).

For consistency and avoiding that scenario (even if errored jobs might behave slightly different), can you move the call to `job.error_message()` or a similar solution that deals with "read-once" results - perhaps populating `self._api_error_msg` only the first time it is called or a similar measure, etc.

diego-plan9 commented on qiskit/providers/ibmq/ibmqjob.py:
```diff
@@ -186,7 +186,21 @@ def result(self, timeout=None, wait=5):
         return Result.from_dict(job_response['qObjectResult'])
 
-    def _wait_for_result(self, timeout=None, wait=5):
+    def _get_error_details(self):
```
Can you add a docstring - and while you are at it, add it also to `_wait_for_result`, taking advantage of the momentum :grin: ? They are the only two methods missing on the class, and it would be great to have all the class documented.

atilag commented on qiskit/providers/ibmq/ibmqjob.py:
```diff
@@ -197,7 +211,7 @@ def _wait_for_result(self, timeout=None, wait=5):
 
         status = self.status()
-        if status is not JobStatus.DONE:
+        if not expect_error and status is not JobStatus.DONE:
```
What about naming `expect_error` as `expect_job_error` or `expect_general_error`? it's more verbose, but can't be misinterpreted with: "I'm expecting errors, even though they are coming from each of the experiments individually"

delapuente replied:
`expect_error` is no longer needed. I think this belongs to an outdated commit.

delapuente commented on qiskit/providers/ibmq/ibmqjob.py:
```diff
@@ -186,7 +186,21 @@ def result(self, timeout=None, wait=5):
+    def _get_error_details(self):
+        error_result = self._wait_for_result(expect_error=True)
+        results = error_result['qObjectResult']['results']
+        if len(results) == 1:
```
I can change this but since this is the usual case in Terra, I preferred to simply return the error message without the report.

diego-plan9 replied:
Actually, I'd go with @atilag's suggestion - seems more homogeneous (non-blocking)

delapuente commented:
@diego-plan9, this is ready for another review.

diego-plan9 commented on qiskit/providers/ibmq/ibmqjob.py:
```diff
@@ -281,7 +303,22 @@ def status(self):
+        Calling this method if the job status is ERROR, consumes the job.
+
+        Returns:
+            str: An error report if the job errored or None elsewhere.
```
Minor nit - `otherwise` instead of `elsewhere`?

diego-plan9 commented on qiskit/providers/ibmq/ibmqjob.py:
```diff
@@ -398,32 +470,22 @@ def _wait_for_submission(self, timeout=60):
-        Note that this method might involve querying the API for results if the
-        Job has been created in a previous Qiskit session.
+    def _build_error_report(self, results):
+        """Build the error report.
```
Can you also document `results`?

diego-plan9 commented on qiskit/providers/ibmq/ibmqjob.py:
```diff
@@ -281,7 +303,22 @@ def status(self):
+
+        job_response = self._get_job()
+        results = job_response['qObjectResult']['results']
+        if not self._api_error_msg:
```
Can you move this `if` block higher (before `_get_job()`) and adjust accordingly (as in, probably including the previous two lines in an `else` or similar)? Otherwise it seems it will attempt to consume the job twice, and the function might error out at that point ungracefully.

diego-plan9 commented on qiskit/providers/ibmq/ibmqjob.py:
```diff
@@ -281,7 +303,22 @@ def status(self):
+        job_response = self._get_job()
+        results = job_response['qObjectResult']['results']
+        if not self._api_error_msg:
+            self._api_error_msg = self._build_error_report(results)
```
Probably minor - but in theory there might be a case where the job errors out at a "general" level, without specific errors in the experiments (as per the `JobStatus` definition), and if that happens it seems the error report will likely fail or not provide meaningful information. Non-blocking as I think in practice it is an unlikely situation, but can you follow up on this case or keep an eye if this happens?

delapuente replied:
I had it on my radar (and I'm still trying to force the situation) but I'll document it in a new issue. :+1:

delapuente commented:
@diego-plan9 ready again!

delapuente commented:
@diego-plan9 I've added the changelog but decided not to expose "consuming the job" concept in the documentation. Instead, I've modified and homogenized ``error_message()`` and ``result()`` to allow more than one calls by caching the result. The interaction with the job states, although complex, it is the expected so I don't think providing extra information would be of any help.

diego-plan9 commented on qiskit/providers/ibmq/ibmqjob.py:
```diff
@@ -220,12 +221,11 @@ def result(self, timeout=None, wait=5):
+            self._result = Result.from_dict(job_response['qObjectResult'])
 
-        return Result.from_dict(job_response['qObjectResult'])
+        return self._result
```
Can you revise the docstring? It has fallen out of sync with your latest comment.
